Draft-Accepted/RFCXXXX-FeedbackProviders.md

```diff
@@ -0,0 +1,192 @@
+
+Feedback providers are triggered on the following:
+- **Comments**
+- **Success**
+- **CommandNotFound**
+- **Error**
+- **All**
+
+The current user interface schema is as follows:
+```
+[Name of first feedback provider]
+  Required Header:
+    ➤ optional action 1
+    ➤ optional action 2
+    ➤ optional action n
+  optional Footer
+...
+[Name of n feedback provider]
+  Required Header:
+    ➤ optional action 1
+    ➤ optional action 2
+    ➤ optional action n
+  optional Footer
+```
+
+> Note: For the sake of collaboration I am including feedback provider UIs in text rather than
+> screenshots. This means the coloring will be off to the final experience.
+
+### On Comment
+
+Comments allow for a more descriptive input for feedback providers to utilize and act upon. One
+example implementation could be using a feedback provider as a pass through to an off box A.I model.
+This way the more detailed input can be used by a large language model.
+
+```powershell
+PS># How do I list the files in my C drive?
+[A.I Model]
+  One way to list all your files in the C drive is the following:
+    ➤ Get-ChildItem -Path C:
+```
+
+### On Success
+
+Feedback providers can trigger on successful executions in order to give suggestions or assistance
+on potentially "better" ways to utilize a command.
+
+One way we hope to use this is with a feedback provider we are
+calling `JSON Adapter` which can help guide users to wrap native commands with adapters that
+convert the text output to a machine readable format or the tool like
+[jc](https://github.com/kellyjonbrazil/jc). This way users can use native commands just like
+PowerShell cmdlets that output an object.
+
+```powershell
+PS> uname
+Darwin
+
+[JSON Adapter]
+  The most similar commands are:
+    ➤ uname | uname-json | ConvertFrom-JSON
+    ➤ uname | jc --uname | ConvertFrom-JSON
+```
+
+### On CommandNotFound
+
+Currently the only trigger on by **_default_** is the CommandNotFound trigger. This triggers specifically
+on the `CommandNotFound` exeception thrown by PowerShell. This specifically targets the scenario for
+mistyping or not having a particular command installed yet.
+
+```powershell
+PS> Get-Itam
+Get-Itam: The term 'Get-Itam' is not recognized as a name of a cmdlet, function, script file, or executable program.
+Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
+
+[general]
+  The most similar commands are:
+    ➤ Get-Item
+```
+
+What this looks like properly rendered is like so:
+
+![Command not found image example](../assets/FeedbackProviders/CommandNotFoundExample.png)
+
+We also developed a feedback provider we are calling the `command-not-found` feedback provider which
+depends on the `command-not-found` utility available on Linux by default. It can be found
+[here](https://github.com/PowerShell/command-not-found). It triggers off the same `CommandNotFound`
+execption but is only compatible with linux systems.
+
+### On Error
```

Comment:
Perhaps we should consider an additional filter so that a provider isn't called for all errors just to discard it.  For example, AzPS Feedback Provider would likely only want to be invoked for an error from a cmdlet from an Az module and also CommandNotFound (like if you made a typo for an Az cmdlet).  However, all other types of errors may simply be causing unnecessary CPU usage.  In this case, it may make sense to specify errors from a set of modules.  Certainly, it makes sense to still allow all errors if such a FeedbackProvider handles many types of cases.

Reply:
I think the experience we are going for is that we will pass the error record object to the feedback provider and then its up to them to trigger on the conditions they set for that object. I.e for your example AzPS could _only_ trigger on `<Verb>-Az*` commands in the error record object. @daxian-dbw can correct me if this is wrong